src/librustc_mir/build/expr/as_place.rs

```diff
@@ -28,22 +28,38 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
         where M: Mirror<'tcx, Output=Expr<'tcx>>
     {
         let expr = self.hir.mirror(expr);
-        self.expr_as_place(block, expr)
+        self.expr_as_place(block, expr, false)
+    }
+
+    /// Compile `expr`, yielding a place that we can move from etc.
+    pub fn as_immutable_place<M>(&mut self,
+                        block: BasicBlock,
+                        expr: M)
+                        -> BlockAnd<Place<'tcx>>
+        where M: Mirror<'tcx, Output=Expr<'tcx>>
+    {
+        let expr = self.hir.mirror(expr);
+        self.expr_as_place(block, expr, true)
     }
 
     fn expr_as_place(&mut self,
                       mut block: BasicBlock,
-                      expr: Expr<'tcx>)
+                      expr: Expr<'tcx>,
+                      immutable: bool)
```

Comment:
maybe we can pass a `hir::Mutability` instead -- something more declarative than a naked boolean